container.go

```diff
@@ -0,0 +1,342 @@
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	return nil
+}
+
+// Wait synchronously waits for the container to shutdown or terminate.
+func (container *hcsShimContainer) Wait() error {
+	_, err := container.waitTimeoutInternal(syscall.INFINITE)
+	return err
+}
+
+// WaitTimeout synchronously waits for the container to terminate or the duration to elapse. It
+// returns true if the event occurs successfully without timeout.
+func (container *hcsShimContainer) WaitTimeout(timeout time.Duration) (bool, error) {
+	// TODO can this overflow uint32
+	return container.waitTimeoutInternal(uint32(timeout.Nanoseconds() / (1000 * 1000)))
+}
+
+func (container *hcsShimContainer) waitTimeoutInternal(timeout uint32) (bool, error) {
+	title := "HCSShim::Container::waitTimeoutInternal"
+	var (
+		resultp   *uint16
+		exitEvent syscall.Handle
+	)
+
+	if err := hcsCreateComputeSystemWait(container.handle, &exitEvent, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return false, err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+
+	s, e := syscall.WaitForSingleObject(syscall.Handle(exitEvent), timeout)
+	switch s {
+	case syscall.WAIT_OBJECT_0:
+		break
+	case syscall.WAIT_FAILED:
+		return false, makeError(e, title, "WaitForSingleObject failed")
+	default:
+		return false, errors.New("os: unexpected result from WaitForSingleObject")
+	}
+
+	return true, nil
+}
+
+func (container *hcsShimContainer) Properties() (*ContainerProperties, error) {
+	title := "HCSShim::Container::Properties"
+	var (
+		resultp     *uint16
+		propertiesp *uint16
+	)
+
+	if err := hcsGetComputeSystemProperties(container.handle, "", &propertiesp, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return nil, err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
+		result := convertAndFreeCoTaskMemString(resultp)
+		logrus.Debugf("%s returned result: %s", title, result)
+	}
+	if propertiesp == nil {
+		return nil, errors.New("Unexpected result from hcsGetComputeSystemProperties, properties should never be nil")
+	}
+	propertiesRaw := convertAndFreeCoTaskMemString(propertiesp)
+
+	properties := &ContainerProperties{}
+	err := json.Unmarshal([]byte(propertiesRaw), properties)
+	if err != nil {
+		return nil, err
+	}
+
+	return properties, nil
+}
+
+// Pause pauses the execution of the container.
+func (container *hcsShimContainer) Pause() error {
+	title := "HCSShim::Container::Pause"
+	var (
+		options string
+		resultp *uint16
+	)
+
+	if err := hcsPauseComputeSystem(container.handle, options, &resultp); err != nil {
+		err = makeError(err, title, "")
+		logrus.Error(err)
+		return err
+	}
+	// result not currently used, but needs to be properly freed.
+	if resultp != nil {
```

Comment:
Make a helper function to handle resultp. Eventually we will allow resultp to influence error, so I suggest making a function:

    func processHcsResult(err error, resultp *uint16) error

For now it would just log resultp and pass err through.